contracts/PublicFundraising.sol

```diff
@@ -187,21 +182,81 @@ contract PublicFundraising is
         tokensSold += _amount;
         tokensBought[_tokenReceiver] += _amount;
 
-        // rounding up to the next whole number. Investor is charged up to one currency bit more in case of a fractional currency bit.
-        uint256 currencyAmount = Math.ceilDiv(_amount * getPrice(), 10 ** token.decimals());
+        token.mint(_tokenReceiver, _amount);
+    }
 
+    function _getFeeAndFeeReceiver(uint256 _currencyAmount) internal view returns (uint256, address) {
+        IFeeSettingsV2 feeSettings = token.feeSettings();
+        return (feeSettings.publicFundraisingFee(_currencyAmount), feeSettings.publicFundraisingFeeCollector());
+    }
+
+    /**
+     * @notice Buy `amount` tokens and mint them to `_tokenReceiver`.
+     * @param _amount amount of tokens to buy, in bits (smallest subunit of token)
+     * @param _tokenReceiver address the tokens should be minted to
+     */
+    function buy(uint256 _amount, address _tokenReceiver) public whenNotPaused nonReentrant {
+        _checkAndDeliver(_amount, _tokenReceiver);
+        // rounding up to the next whole number. Investor is charged up to one currency bit more in case of a fractional currency bit.
+        uint256 currencyAmount = calculateCurrencyAmountFromTokenAmount(_amount);
         IFeeSettingsV2 feeSettings = token.feeSettings();
         uint256 fee = feeSettings.publicFundraisingFee(currencyAmount);
         if (fee != 0) {
             currency.safeTransferFrom(_msgSender(), feeSettings.publicFundraisingFeeCollector(), fee);
         }
 
         currency.safeTransferFrom(_msgSender(), currencyReceiver, currencyAmount - fee);
-
-        token.mint(_tokenReceiver, _amount);
         emit TokensBought(_msgSender(), _amount, currencyAmount);
     }
 
+    /// calculate token amount from currency amount and price. Must be rounded down anyway, so the normal integer math is fine.
+    /// This calculation often results in a larger amount of tokens
+    function calculateTokenAmountFromCurrencyAmount(uint256 _currencyAmount) public view returns (uint256) {
+        return (_currencyAmount * 10 ** token.decimals()) / getPrice();
+    }
+
+    function calculateCurrencyAmountFromTokenAmount(uint256 _tokenAmount) public view returns (uint256) {
+        return Math.ceilDiv(_tokenAmount * getPrice(), 10 ** token.decimals());
+    }
+
+    function findMaxAmount(uint256 _minAmount) external view returns (uint256) {
```

Comment:
Why do we have this function? It is never used (except tests)

Reply:
I imagine ourselves and project wanting to integrate with us wanting to use this function. It answers the question "If I want to buy X amount of tokens, which higher amount Y can I buy without paying more?", which I personally would always be interested in.
Another approach would be to expose more buy functions:
- buyExact
- buyAsMuchAsPossible

I chose this view function instead because view functions are less dangerous, but can be nested to achieve almost the same result.